### PR TITLE
Build troubleshooting for long builds

### PR DESCRIPTION
**Pull request by AMZN-alexpete**
## Change summary

Troubleshooting steps for long builds or builds that freeze or lock up a computer with high CPU core counts

### Submission Checklist:

* [x] **Descriptive active voice** - Do descriptive sentences have a clear *subject* and *action verb*?
* [x] **Answer the question at hand** - Does the documentation answer a *what*, *why*, *how*, or *where* type of question?
* [x] **Consistency** - Does the content consistently follow the [Style Guide](https://o3de.org/docs/contributing/to-docs/style-guide/quick-reference)?
* [x] **Help the user** - Does the documentation show the user something *meaningful*?

### CODE REVIEW

**Comment by AMZN-alexpete** [comment]
Considering adding an image showing VS 2022's GUI for setting Maximum number of parallel builds 
image
and maximum concurrent c++ compilations
image

**Review by lumberyard-employee-dm — APPROVED** [review]
We should also mention on Linux that the build could run out of memory and cause OOM killer to run or for certain install cause the machine to reboot.

The cmake --build `--j/--parallel` option can be used to reduce the number of parallel.

For example on a 12-core 32-GiB machine, the user might want to reduce the number of parallel build jobs to 6 if they are noticing their compile processes are being killed.
```bash
cmake --build <build-dir> --j 6 --config profile --target MyProject.GameLauncher -- <generator specific options>
```

**Comment by AMZN-alexpete** [comment]
Went ahead and added those VS 2022 menu images, suggestion to use unity builds, and cmake parallel build option.

File: content/docs/user-guide/build/troubleshooting.md [hunk 1]
```diff
@@ -130,20 +130,32 @@ time cmake --build build/linux --target Editor AutomatedTesting.GameLauncher --c
 1. Make sure you are only building the targets you are using.  For example, select an appropriate target like the `Editor` for building and not the `ALL_BUILD` target which will build everything and take longer.
   Keep in mind that your first build will take the longest because all the code for the selected target and dependencies must be compiled, but subsequent builds should be much faster (minutes or seconds) especially when writing game code.
 1. Use the `profile` build configuration when developing when you don't need all the un-optimized debugging information of a `debug` build configuration.
+1. Make sure you're using unity builds (on by default) by configuring with the option `-DLY_UNITY_BUILD=ON` build configuration when developing when you don't need all the un-optimized debugging information of a `debug` build configuration.
```
Comment (nicholas-rh) [inline content/docs/user-guide/build/troubleshooting.md:133]: It looks like the "build configuration when developing when you don't need all the un-optimized debugging information of a `debug` build configuration" blurb tagged along from the previous item in the list here

Comment (AMZN-alexpete) [reply content/docs/user-guide/build/troubleshooting.md]: good catch!

**Review by lumberyard-employee-dm — APPROVED** [review]
Looks good to me

**Review by AMZN-daimini — APPROVED** [review]

**Comment by AMZN-alexpete** [comment]
Added @willihay to the review

File: content/docs/user-guide/build/troubleshooting.md [hunk 2]
```diff
@@ -104,3 +104,89 @@ Call Stack (most recent call first):
...
+## Building causes computer to freeze, lock up or reboot
+
+**Issue:** On some computers, especially those with CPUs that have high core counts, the way compilation is spread across cores can lead to resource starvation and even freezing.  It's also possible that a build could run out of memory and cause issues including resource starvation and reboot.
+
+**Remedy:** Adjust the build settings to spread out the maximum number of parallel processes used and the maximum CL processor counts allowed per project if possible.
+
+{{< tabs name="Platform Specific Settings" >}}
+{{% tab name="Windows" %}}
+
+For Visual Studio on Windows, the command might look like this to set the maximum cpu count to 16 and limit to 32 CL processor counts per project.
```
Comment (willihay) [inline content/docs/user-guide/build/troubleshooting.md:147]: Adjust word order and plurality for continuity.
```suggestion
For Visual Studio on Windows, the command might look like this to set the maximum CPU count to 16 and limit the CL processor count to 32 per project.
```

File: content/docs/user-guide/build/troubleshooting.md [hunk 3]
```diff
@@ -104,3 +104,89 @@ Call Stack (most recent call first):
...
+## Long build times
+
+**Issue:** Building takes longer than expected.
+
+**Remedy:** There are several steps to troubleshooting builds that are taking a long time.
+1. Use a build command that measures the amount of time taken so you can compare build times when troubleshooting.  On Windows you can use the PowerShell `measure-command` to measure build times, and on linux `time`.
+{{< tabs name="CMake build timing example" >}}
+{{% tab name="Windows" %}}
+
+```cmd
```
Comment (willihay) [inline content/docs/user-guide/build/troubleshooting.md:117]: For better PowerShell syntax highlighting, use `pwsh` in all your PowerShell examples.
```suggestion
```pwsh
```

File: content/docs/user-guide/build/troubleshooting.md [hunk 4]
```diff
@@ -104,3 +104,89 @@ Call Stack (most recent call first):
...
+
+**Issue:** On some computers, especially those with CPUs that have high core counts, the way compilation is spread across cores can lead to resource starvation and even freezing.  It's also possible that a build could run out of memory and cause issues including resource starvation and reboot.
+
+**Remedy:** Adjust the build settings to spread out the maximum number of parallel processes used and the maximum CL processor counts allowed per project if possible.
+
+{{< tabs name="Platform Specific Settings" >}}
+{{% tab name="Windows" %}}
+
+For Visual Studio on Windows, the command might look like this to set the maximum cpu count to 16 and limit to 32 CL processor counts per project.
+```cmd
```
Comment (willihay) [inline content/docs/user-guide/build/troubleshooting.md:148]:
```suggestion
```pwsh
```

File: content/docs/user-guide/build/troubleshooting.md [hunk 5]
```diff
@@ -104,3 +104,89 @@ Call Stack (most recent call first):
...
+![Visual Studio Maximum Number of Parallel Build Options](/images/user-guide/build/visual-studio-options-build-and-run.jpg)
+
+Set the Maximum Concurrent C++ Compilations in the Tools -> Options -> Projects and Solutions -> VC++ Project Settings option page:
+
+![Visual Studio Maximum Concurrent C++ Compilations](/images/user-guide/build/visual-studio-options-vc++-project-settings.jpg)
+
+Experiment with different values to find the best settings for your hardware.  
+On Windows you can measure your compile times in a PowerShell terminal with different settings with a command like this:
+
+```cmd
```
Comment (willihay) [inline content/docs/user-guide/build/troubleshooting.md:162]:
```suggestion
```pwsh
```

File: content/docs/user-guide/build/troubleshooting.md [hunk 6]
```diff
@@ -104,3 +104,89 @@ Call Stack (most recent call first):
...
+TotalHours        : 0.265750398722222
+TotalMinutes      : 15.9450239233333
+TotalSeconds      : 956.7014354
+TotalMilliseconds : 956701.4354
+```
+
+{{% /tab %}}
+{{% tab name="Linux" %}}
+
+Use the [`-j/--parallel` cmake option](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-build-j) to reduce the number of parallel jobs used when building.
```
Comment (willihay) [inline content/docs/user-guide/build/troubleshooting.md:183]: The either/or option is a little confusing when rendered in a hyperlink.
```suggestion
Use the [`-j` or `--parallel` cmake option](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-build-j) to limit the number of parallel jobs used when building.
```

**Review by willihay — APPROVED** [review]

**Pull request merged**